server/src/main/java/org/elasticsearch/gateway/TransportNodesListGatewayStartedShards.java

```diff
@@ -111,13 +115,30 @@ protected NodesGatewayStartedShards newResponse(Request request,
         return new NodesGatewayStartedShards(clusterService.getClusterName(), responses, failures);
     }
 
+    private ShardStateMetaData safelyLoadLatestState(ShardId shardId) throws Exception {
+        final IndexShard indexShard = indicesService.getShardOrNull(shardId);
+        if (indexShard != null) {
+            try {
+                return indexShard.loadShardStateMetaDataIfOpen(NamedXContentRegistry.EMPTY, nodeEnv.availableShardPaths(shardId));
+            }
+            catch (AlreadyClosedException ignored) {
+                // Ok, fall through to trying to get the shard lock ourselves.
+            }
+        }
+
+        try (ShardLock ignored = nodeEnv.shardLock(shardId, TimeUnit.SECONDS.toMillis(5))) {
+            return ShardStateMetaData.FORMAT.loadLatestState(logger, NamedXContentRegistry.EMPTY,
+                nodeEnv.availableShardPaths(shardId));
+        }
+    }
+
     @Override
     protected NodeGatewayStartedShards nodeOperation(NodeRequest request) {
         try {
             final ShardId shardId = request.getShardId();
             logger.trace("{} loading local shard state info", shardId);
-            ShardStateMetaData shardStateMetaData = ShardStateMetaData.FORMAT.loadLatestState(logger, NamedXContentRegistry.EMPTY,
-                nodeEnv.availableShardPaths(request.shardId));
+
+            ShardStateMetaData shardStateMetaData = safelyLoadLatestState(shardId);
```

Comment:
It's nice this logic was folded into a method, but I think we do the shard level resolving here the whole thing will be simpler and we won't need to resolve the index shard twice?

Reply:
Ok, I moved this code around in 7f835ccb460. I'm not 100% comfortable with the changes made since I'm unfamiliar with all the invariants that may or may not hold here - please tread carefully.